k2/csrc/array.h

```diff
@@ -410,10 +417,46 @@ class Array2 {
     that
      it will have a different memory layout than the input.
   */
-  Array2<T> To(ContextPtr ctx) {
-    // TODO
-    Array2<T> array;
-    return array;
+  Array2<T> To(ContextPtr ctx) const {
+    if (dim0_ * dim1_ == 0) {
+      if (!region_) return *this;
+
+      if (ctx->IsCompatible(*Context())) return *this;
+
+      Array2 ans(ctx, dim0_, dim1_);
+      return ans;
+    }
+
+    if (ctx->IsCompatible(*Context()) && elem_stride0_ == dim1_) return *this;
+
+    // the current array is either non-contiguous or not compatible with the ctx
+
+    Array2<T> ans(ctx, dim0_, dim1_);
+
+    if (elem_stride0_ == dim1_) {
+      // the current array is contiguous, use memcpy
+      auto kind = GetMemoryCopyKind(*Context(), *ctx);
+      auto *dst = ans.Data();
+      const auto *src = Data();
+      MemoryCopy(static_cast<void *>(dst), static_cast<const void *>(src),
+                 dim0_ * dim1_ * ElementSize(), kind);
+      return ans;
+    } else {
+      // do rowwise copy
+      auto *dst = ans.Data();
+      const auto *src = Data();
+
+      auto kind = GetMemoryCopyKind(*Context(), *ctx);
+
+      for (int32_t i = 0; i != dim0_; ++i) {
+        MemoryCopy(static_cast<void *>(dst), static_cast<const void *>(src),
```

Comment:
it could be done with kernel, see my implementation in   `Flatten`

Reply:
The source context and dest context may represent different devices, e.g., one for CPU and the other for GPU.
I am afraid in this case, we cannot use a kernel to replace memory copy.

Reply:
yeah, kernel using Eval2(), a 2d kernel.

Reply:
.. eventually we can switch to a more specialized implementation that used the hardware in a smarter way.

Reply:
And for non-compatiable context, it may be better to make it contiguous first with `kernel` in a same Context, and then do a one pass copy.

Reply:
@csukuangfj if the source array was not contiguous, we first make it contiguous *on the same device* by calling ToContiguous().  We need to define this function, e.g. can put it in ops.h:  Array2<T> ToContiguous(const Array2<T> &src);

Reply:
.. please see ToContiguous() in tensor_ops.h for reference.

Reply:
fixed. But I implemented it as a member function.

Reply:
Moved to ops.h and ops_inl.h